src/lib.rs

```diff
@@ -11,6 +11,898 @@
+                    UP_CLIENT_VSOMEIP_TAG,
+                    UP_CLIENT_VSOMEIP_FN_TAG_REGISTER_LISTENER_INTERNAL,
+                );
+                let Some(sink_filter) = sink_filter else {
+                    return Err(UStatus::fail_with_code(
+                        UCode::INVALID_ARGUMENT,
+                        "Unable to map source and sink filters to uProtocol message type",
+                    ));
+                };
+
+                let (_, service_id) = split_u32_to_u16(sink_filter.ue_id);
+                let instance_id = 1; // TODO: Set this to 1? To ANY_INSTANCE?
+                let (_, method_id) = split_u32_to_u16(sink_filter.resource_id);
+
+                trace!(
+                    "{}:{} - register_message_handler: service: {} instance: {} method: {}",
+                    UP_CLIENT_VSOMEIP_TAG,
+                    UP_CLIENT_VSOMEIP_FN_TAG_REGISTER_LISTENER_INTERNAL,
+                    service_id,
+                    instance_id,
+                    method_id
+                );
+
+                {
+                    let mut offered_services = OFFERED_SERVICES.write().await;
+                    if !offered_services.contains(&(service_id, instance_id, method_id)) {
+                        get_pinned_application(application_wrapper).offer_service(
+                            service_id,
+                            instance_id,
+                            ANY_MAJOR,
+                            ANY_MINOR,
+                        );
+                        offered_services.insert((service_id, instance_id, method_id));
+                    }
+                }
+
+                register_message_handler_fn_ptr_safe(
+                    application_wrapper,
+                    service_id,
+                    vsomeip::ANY_INSTANCE,
+                    method_id,
+                    msg_handler,
+                );
+
+                trace!(
+                    "{}:{} - Registered vsomeip message handler.",
+                    UP_CLIENT_VSOMEIP_TAG,
+                    UP_CLIENT_VSOMEIP_FN_TAG_REGISTER_LISTENER_INTERNAL,
+                );
+
+                Ok(())
+            }
+            RegistrationType::Response(_) => {
+                trace!(
+                    "{}:{} - Registering for Response style messages.",
+                    UP_CLIENT_VSOMEIP_TAG,
+                    UP_CLIENT_VSOMEIP_FN_TAG_REGISTER_LISTENER_INTERNAL,
+                );
+
+                let (_, service_id) = split_u32_to_u16(source_filter.ue_id);
+                let instance_id = vsomeip::ANY_INSTANCE; // TODO: Set this to 1? To ANY_INSTANCE?
+                let (_, method_id) = split_u32_to_u16(source_filter.resource_id);
+
+                {
+                    let requested_services = REQUESTED_SERVICES.write().await;
+                    if !requested_services.contains(&(service_id, instance_id, method_id)) {
+                        get_pinned_application(application_wrapper).request_service(
```

Comment:
You don't need to call `request_service()` multiple times for the same service if it has multiple methods